cdc/puller/multiplexing_puller.go

```diff
@@ -0,0 +1,428 @@
+	frontiers int,
+) *MultiplexingPuller {
+	x := &MultiplexingPuller{
+		changefeed: changefeed,
+		client:     client,
+		consume:    consume,
+		workers:    workers,
+		hasher:     hasher,
+		frontiers:  frontiers,
+		advanceCh:  make(chan *tableProgress, 128),
+	}
+	x.subscriptions.m = make(map[kv.SubscriptionID]*tableProgress)
+	x.subscriptions.n = spanz.NewHashMap[*tableProgress]()
+
+	x.inputChs = make([]chan kv.MultiplexingEvent, 0, workers)
+	for i := 0; i < workers; i++ {
+		x.inputChs = append(x.inputChs, make(chan kv.MultiplexingEvent, 1024))
+	}
+	return x
+}
+
+// Subscribe some spans. They will share one same resolved timestamp progress.
+func (p *MultiplexingPuller) Subscribe(spans []tablepb.Span, startTs model.Ts, tableName string) {
+	p.subscriptions.Lock()
+	defer p.subscriptions.Unlock()
+	p.subscribe(spans, startTs, tableName)
+}
+
+func (p *MultiplexingPuller) subscribe(spans []tablepb.Span, startTs model.Ts, tableName string) []kv.SubscriptionID {
+	progress := &tableProgress{
+		changefeed: p.changefeed,
+		client:     p.client,
+		spans:      spans,
+		startTs:    startTs,
+		tableName:  tableName,
+
+		resolvedSpans: make(chan kv.MultiplexingEvent, 16),
+		tsTracker:     frontier.NewFrontier(0, spans...),
+	}
+
+	progress.consume.f = func(ctx context.Context, raw *model.RawKVEntry, spans []tablepb.Span) error {
+		progress.consume.RLock()
+		defer progress.consume.RUnlock()
+		if !progress.consume.removed {
+			return p.consume(ctx, raw, spans)
+		}
+		return nil
+	}
+
+	subIDs := make([]kv.SubscriptionID, len(spans))
+	for i, span := range spans {
+		subID := p.client.AllocSubscriptionID()
+		subIDs[i] = subID
+
+		p.subscriptions.m[subID] = progress
+		p.subscriptions.n.ReplaceOrInsert(span, progress)
+
+		slot := p.hasher(span, len(p.inputChs))
+		if _, ok := p.client.Subscribe(subID, span, startTs, p.inputChs[slot]); !ok {
+			log.Panic("redundant subscription",
+				zap.String("namespace", p.changefeed.Namespace),
+				zap.String("changefeed", p.changefeed.ID),
+				zap.String("span", span.String()))
+		}
+	}
+	return subIDs
+}
+
+// Unsubscribe some spans, which must be subscribed in one call.
+func (p *MultiplexingPuller) Unsubscribe(spans []tablepb.Span) {
+	p.subscriptions.Lock()
+	defer p.subscriptions.Unlock()
+	p.unsubscribe(spans)
+}
+
+func (p *MultiplexingPuller) unsubscribe(spans []tablepb.Span) {
+	for _, span := range spans {
+		if subID, ok := p.client.Unsubscribe(span); ok {
+			progress := p.subscriptions.m[subID]
+			progress.consume.Lock()
+			progress.consume.removed = true
+			progress.consume.Unlock()
+
+			delete(p.subscriptions.m, subID)
+			p.subscriptions.n.Delete(span)
+		} else {
+			log.Panic("unexist unsubscription",
+				zap.String("namespace", p.changefeed.Namespace),
+				zap.String("changefeed", p.changefeed.ID),
+				zap.String("span", span.String()))
+		}
+	}
+}
+
+// Run the puller.
+func (p *MultiplexingPuller) Run(ctx context.Context) (err error) {
```

Comment:
Is it necessary to print a log when exiting?

Reply:
yes, it's reasonable.